File: SampleProjects/GoT-Characters-master/Objective-C/GoT/MainViewController.m
//
//  MainViewController.m
//  GoT
//
//  Created by Paciej on 21/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

#import "MainViewController.h"
#import "DetailsViewController.h"
#import "FavouriteTableViewCell.h"
#import "Article.h"
#import "Configurator.h"

#import <Masonry/Masonry.h>

static NSString *kFavourite = @"Favourite";
static NSString *kAll = @"All";

@interface MainViewController () <UITableViewDelegate>

@property(nonatomic, strong, nonnull) Loader *loader;
@property(nonatomic, strong, nonnull) AsyncLoadConfiguration *configuration;
@property(nonatomic, strong, nonnull) DataSource *dataSource;
@property(nonatomic, strong, nonnull) ArticlesRepository *articlesRepository;
@property(nonatomic, strong, nonnull) UITableView *tableView;
@property(nonatomic, strong, nonnull) NSArray *tableViewDataSource;
@property(nonatomic) BOOL filterEnabled;
@end

@implementation MainViewController

- (nonnull instancetype)
initWithLoader:(nonnull Loader *)loader
asyncLoadConfiguration:(nonnull AsyncLoadConfiguration *)configuration
dataSource:(nonnull DataSource *)dataSource
articlesRepository:(nonnull ArticlesRepository *)articlesRepository {
    self = [super initWithNibName:nil bundle:nil];
    if (self) {
        self.loader = loader;
        self.configuration = configuration;
        self.dataSource = dataSource;
        self.articlesRepository = articlesRepository;
        self.title = @"Characters";
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]
                                                  initWithTitle:kFavourite
                                                  style:UIBarButtonItemStylePlain
                                                  target:self
                                                  action:@selector(rightBarButtonItemTapped:)];
        self.filterEnabled = false;
    }
    return self;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadTableView];
    [self loadArticlesFromWebservice];
}

- (void)loadTableView {
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        self.tableView = [[UITableView alloc] initWithFrame:CGRectZero
                                                      style:UITableViewStylePlain];
        self.tableView.dataSource = self.dataSource;
        self.tableView.delegate = self;
        [self.tableView registerClass:[FavouriteTableViewCell class]
               forCellReuseIdentifier:self.dataSource.cellReuseIdentifier];
        [self.view addSubview:self.tableView];
        [self.tableView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.edges.equalTo(self.tableView.superview);
        }];
        self.dataSource.tableView = self.tableView;
        __weak __typeof__(self) __weakSelf = self;
        self.dataSource.cellConfigureBlock = ^(UITableViewCell *_Nonnull cell,
                                               NSIndexPath *_Nonnull indexPath,
                                               id _Nonnull item) {
            FavouriteTableViewCell *favCell = (FavouriteTableViewCell *)cell;
            if ([item isKindOfClass:[Article class]]) {
                Article *article = (Article *)item;
                favCell.titleLabel.text = article.title;
                favCell.abstractLabel.text = article.abstract;
                favCell.tag = indexPath.row;
                favCell.favouriteButton.tag = indexPath.row;
                if (article.thumbnailData) {
                    favCell.imageView.image = [article imageFromThumbnailData];
                }
                __strong __typeof__(__weakSelf) strongSelf = __weakSelf;
                if (strongSelf) {
                    [favCell.favouriteButton addTarget:strongSelf
                                                action:@selector(favouriteButtonTapped:)
                                      forControlEvents:UIControlEventTouchUpInside];
                    favCell.favouriteButton.selected =
                    [strongSelf.articlesRepository isFavouriteArticle:article];
                    ;
                }
            }
        };
    });
}

- (void)loadArticlesFromWebservice {
    [self.loader
     loadAsynchronously:self.configuration
     callback:^(id _Nullable result) {
         if ([result isKindOfClass:[NSArray class]]) {
             [self.articlesRepository storeArticlesTemporarily:result];
             [self.dataSource
              addItems:result]; // reloads table view on main queue
         }
     }];
}

- (CGFloat)tableView:(UITableView *)tableView
heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 100.0;
}

- (void)tableView:(UITableView *)tableView
  willDisplayCell:(UITableViewCell *)cell
forRowAtIndexPath:(NSIndexPath *)indexPath {
    __block Article *article = [self.dataSource itemAtIndexPath:indexPath];
    
    if (!article.thumbnailData) {
        __block FavouriteTableViewCell *favouriteCell;
        __block NSInteger tag = indexPath.row;
        if ([cell isKindOfClass:[FavouriteTableViewCell class]]) {
            favouriteCell = (FavouriteTableViewCell *)cell;
        }
        AsyncLoadConfiguration *configuration =
        [AsyncLoadConfiguration asyncLoadConfigurationFromArticle:article];
        [self.loader loadAsynchronously:configuration
                               callback:^(id _Nullable result) {
                                   if ([result isKindOfClass:[NSData class]]) {
                                       article.thumbnailData = result;
                                       dispatch_async(dispatch_get_main_queue(), ^{
                                            if (favouriteCell.tag == tag) {
                                                favouriteCell.imageView.image =
                                                [article imageFromThumbnailData];
                                                [favouriteCell setNeedsDisplay];
                                            }
                                       });
                                   }
                               }];
    }
}

- (void)tableView:(UITableView *)tableView
didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    Article *article = [self.dataSource itemAtIndexPath:indexPath];
    BOOL favourite = [self.articlesRepository isFavouriteArticle:article];
    DetailsViewController *viewController =
    [Configurator configuredDetailsViewControllerWithArticle:article
                                                   favourite:favourite];
    [self showViewController:viewController sender:self];
}

- (void)favouriteButtonTapped:(UIButton *)sender {
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:sender.tag inSection:0];
    Article *article = [self.dataSource itemAtIndexPath:indexPath];
    if ([self.articlesRepository isFavouriteArticle:article]) {
        [self.articlesRepository removeFavouriteArticle:article];
    } else {
        [self.articlesRepository saveFavouriteArticle:article];
    }
    sender.selected = !sender.selected;
    [sender setNeedsDisplay];
}

- (void)rightBarButtonItemTapped:(UIBarButtonItem *)sender {
    self.filterEnabled = !self.filterEnabled;
    sender.title = (self.filterEnabled) ? kAll : kFavourite;
    if (self.filterEnabled) {
        [self.dataSource
         setItems:[self.articlesRepository savedFavouritesArticles]];
    } else {
        [self.dataSource
         setItems:[self.articlesRepository temporarilyStoredArticles]];
    }
}

@end

File: SampleProjects/GoT-Characters-master/Swift/GoT/MainViewController.swift
//
//  MainViewController.swift
//  GoT
//
//  Created by Paciej on 21/10/15.
//  Copyright © 2015 Maciej Piotrowski. All rights reserved.
//

import Masonry
import UIKit

private var kFavourite = "Favourite"
private var kAll = "All"

class MainViewController: UIViewController, UITableViewDelegate {
    private var loader: Loader!
    private var configuration: AsyncLoadConfiguration!
    private var dataSource: DataSource!
    private var articlesRepository: ArticlesRepository!
    private var tableView: UITableView!
    private var tableViewDataSource: [Any] = []
    private var filterEnabled = false

    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }

    init(loader: Loader, asyncLoadConfiguration configuration: AsyncLoadConfiguration, dataSource: DataSource, articlesRepository: ArticlesRepository) {
        super.init(nibName: nil, bundle: nil)
        
        self.loader = loader
        self.configuration = configuration
        self.dataSource = dataSource
        self.articlesRepository = articlesRepository
        title = "Characters"
        navigationItem.rightBarButtonItem = UIBarButtonItem(title: kFavourite, style: .plain, target: self, action: #selector(MainViewController.rightBarButtonItemTapped(_:)))
        filterEnabled = false
    
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        loadTableView()
        loadArticlesFromWebservice()
    }

    func loadTableView() {
        var onceToken: Int = 0
        // TODO: use a static variable (`dispatch_once` is deprecated)
    if (onceToken == 0) {
            self.tableView = UITableView(frame: CGRect.zero, style: .plain)
            self.tableView.dataSource = self.dataSource
            self.tableView.delegate = self
            self.tableView.register(FavouriteTableViewCell.self, forCellReuseIdentifier: self.dataSource.cellReuseIdentifier)
            self.view.addSubview(self.tableView)
            self.tableView.mas_makeConstraints({ make in
                _ = make?.edges.equalTo()(self.tableView.superview)
            })
            self.dataSource.tableView = self.tableView
            weak var __weakSelf: MainViewController? = self
            self.dataSource.cellConfigureBlock = { cell, indexPath, item in
                let favCell = cell as? FavouriteTableViewCell
                if (item is Article) {
                    let article = item as? Article
                    favCell?.titleLabel.text = article?.title
                    favCell?.abstractLabel.text = article?.abstract
                    favCell?.tag = indexPath.row
                    favCell?.favouriteButton.tag = indexPath.row
                    if article?.thumbnailData != nil {
                        favCell?.imageView?.image = article?.imageFromThumbnailData()
                    }
                    let strongSelf: MainViewController? = __weakSelf
                    if strongSelf != nil {
                        favCell?.favouriteButton.addTarget(strongSelf, action: #selector(MainViewController.favouriteButtonTapped(_:)), for: .touchUpInside)
                        if let anArticle = article {
                            favCell?.favouriteButton.isSelected = strongSelf?.articlesRepository.isFavouriteArticle(anArticle) ?? false
                        }
                    }
                }
            }
        }
    onceToken = 1
    }

    func loadArticlesFromWebservice() {
        loader.loadAsynchronously(configuration, callback: { result in
            if (result is [Any]) {
                if let aResult = result as? [Any] {
                    self.articlesRepository.storeArticlesTemporarily(aResult)
                }
                if let aResult = result as? [Any] {
                    self.dataSource.addItems(aResult)
                } // reloads table view on main queue
            }
        })
    }

    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 100.0
    }

    func tableView(_ tableView: UITableView, willDisplay cell: UITableViewCell, forRowAt indexPath: IndexPath) {
        let article = dataSource.item(at: indexPath) as? Article

        if article?.thumbnailData == nil {
            var favouriteCell: FavouriteTableViewCell?
            let tag: Int = indexPath.row
            if (cell is FavouriteTableViewCell) {
                favouriteCell = cell as? FavouriteTableViewCell
            }
            var configuration: AsyncLoadConfiguration? = nil
            if let anArticle = article {
                configuration = AsyncLoadConfiguration(fromArticle: anArticle) as? AsyncLoadConfiguration
            }
            if let aConfiguration = configuration {
                loader.loadAsynchronously(aConfiguration, callback: { result in
                    if (result is Data) {
                        article?.thumbnailData = result as? Data
                        DispatchQueue.main.async(execute: {
                            if favouriteCell?.tag == tag {
                                favouriteCell?.imageView?.image = article?.imageFromThumbnailData()
                                favouriteCell?.setNeedsDisplay()
                            }
                        })
                    }
                })
            }
        }
    }

    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        let article = dataSource.item(at: indexPath) as? Article
        var favourite: Bool? = nil
        if let anArticle = article {
            favourite = articlesRepository.isFavouriteArticle(anArticle)
        }
        var viewController: DetailsViewController? = nil
        if let anArticle = article {
            viewController = Configurator.configuredDetailsViewController(with: anArticle, favourite: favourite ?? false)
        }
        if let aController = viewController {
            show(aController, sender: self)
        }
    }

    @objc func favouriteButtonTapped(_ sender: UIButton?) {
        let indexPath = IndexPath(row: sender?.tag ?? 0, section: 0)
        let article = dataSource.item(at: indexPath) as? Article
        if let anArticle = article {
            if articlesRepository.isFavouriteArticle(anArticle) {
                articlesRepository.removeFavouriteArticle(anArticle)
            } else {
                articlesRepository.saveFavouriteArticle(anArticle)
            }
        }
        sender?.isSelected = !(sender?.isSelected ?? false)
        sender?.setNeedsDisplay()
    }

    @objc func rightBarButtonItemTapped(_ sender: UIBarButtonItem?) {
        filterEnabled = !filterEnabled
        sender?.title = (filterEnabled) ? kAll : kFavourite
        if filterEnabled {
            dataSource.setItems(articlesRepository.savedFavouritesArticles())
        } else {
            dataSource.setItems(articlesRepository.temporarilyStoredArticles())
        }
    }
}